shell/run_shell_cmd.go

```diff
@@ -137,3 +166,43 @@ func (signalChannel *SignalsForwarder) Close() error {
 	close(*signalChannel)
 	return nil
 }
+
+// This function captures stdout and stderr while still printing it to the stdout and stderr of this Go program
+// outToErr flips std out to std err
+func readStdoutAndStderr(stdout io.ReadCloser, stderr io.ReadCloser, terragruntOptions *options.TerragruntOptions, outToErr bool) (string, error) {
+	allOutput := []string{}
+
+	var errWriter = terragruntOptions.ErrWriter
+	var outWriter = terragruntOptions.Writer
+
+	if outToErr {
+		outWriter = terragruntOptions.ErrWriter
+	}
+
+	stdoutScanner := bufio.NewScanner(stdout)
+	stderrScanner := bufio.NewScanner(stderr)
+
+	for {
+		if stdoutScanner.Scan() {
+			text := stdoutScanner.Text()
+			fmt.Fprintln(outWriter, text)
+			allOutput = append(allOutput, text)
+		} else if stderrScanner.Scan() {
+			text := stderrScanner.Text()
+			fmt.Fprintln(errWriter, text)
+			allOutput = append(allOutput, text)
+		} else {
+			break
+		}
+	}
```

Comment:
Cool, thanks!

Reply:
Hm, won't this change how the output looks? If the service was logging intermittently to stdout and stderr, I think the code as written above would write everything in stdout first and then everything in stderr after. We should add a unit test specifically for this method if we can where we execute some code with known (interleaved) stdout and stderr and make sure we get back *exactly* what we expect (and not just string contains).

Reply:
You might be right. I'll look into it and write a unit test.

Reply:
I'm not sure either approach actually work. In the original if statement, the the stdoutScanner is always evaluated first and we never reach the stderrScanner until stdout is drained.

Reply:
I think I am in agreement with @autero1 . I had suggested the above method because I observed the equivalence of outputting stdout before stderr and thought that was the intended behavior, but I had not realized the desired behavior was the interleaving.